src/shared/tree_container.h

```diff
@@ -484,53 +512,154 @@ class TreeContainer
   // a valid block, so all the names of the dead shared memory blocks are kept
   // in a vector and deallocated at the very end
   //
-  void reassign() const
+  void reassign()
   {
-    // safe const_cast, TreeContainer are never created const
-    auto *self = const_cast<TreeContainer<TreeT>*>(this);
-
-    // reassign() was called because the block is full, in which case a new one
-    // will be created and the current one becomes outdated
-    //
-    // reassign() can also be called because `outdated` was already true
-    self->m_TreeMeta->outdated = true;
-
     // list of all the shared memory blocks that are now unused and can be
     // destroyed
     std::vector<std::string> deadSHMNames;
 
-    for (;;) {
+    if (m_TreeMeta->outdated) {
+      // this block was marked as outdated, which should only happen when
+      // another process has ran out of memory and started allocating blocks
+      // with higher numbers
+
+      spdlog::get("usvfs")->info(
+        "tree {0} is outdated, looking for another one", m_SHMName);
+
+      if (findNewerBlock(deadSHMNames)) {
+        // the new block was found and activated
+        return;
+      }
+    } else {
+      // this block isn't outdated, so reassign() was called because a bad_alloc
+      // exception was thrown
+      spdlog::get("usvfs")->info(
+        "ran out of memory in tree {0}, will create another one", m_SHMName);
+    }
+
+    // either the block is full or it's outdated, but no higher block was found;
+    // just create a new one
+
+    createNewBlock(deadSHMNames);
+
+    // remove the old shared memory blocks; this can be recursive and call
+    // reassign() again, but it's safe at this point
+    for (const std::string& name : deadSHMNames) {
+      spdlog::get("usvfs")->info("destroying {0}", name);
+      bi::shared_memory_object::remove(name.c_str());
+    }
+  }
+
+  // tries a series of shm names above the current one in the hope of finding
+  // the most recent one
+  //
+  // if the new block was found and activated correctly, returns true
+  //
+  // when findNewerBlock() returns false, this container will either still be
+  // pointing to the same, outdated block as it started with, or it might have
+  // moved up to another outdated block, but it will always be pointing to
+  // a valid block in memory
+  //
+  bool findNewerBlock(std::vector<std::string>& deadSHMNames)
+  {
+    // how many blocks are checked above the current one; it's unlikely there
+    // will ever be more than 15 to 20 blocks allocated, but this is high
+    // because it's really the only way to keep processes connected to the same
+    // shm block, so processes must not fail to find the next one
+    constexpr int Tries = 100;
```

Comment:
Assuming each shm is double the size of the previous this becomes `(original size) * 2^100` .
Considering 2^40 is a Tera I think we can stop at 50 tries and be pretty confident we got it.

Reply:
Meh, anything above 20 won't happen, but what does it matter?